SCF/OOP_Attempt/Epack.py

```diff
@@ -0,0 +1,284 @@
+#SCF Class definitions
+import time
+import numpy as np
+import psi4
+import scipy.linalg as sl
+from JK_Builder import JK_Builder
+
+class Epack:
+    
+    def __init__(self, wfn, dE_tol, dRMS_tol, maxmacro, SOSCF_tol, maxmicro):#param for SCF iterations, energy types specified below.
+        self.psi4 = psi4
+        self.wfn = wfn
+        self.dE_tol = dE_tol
+        self.maxmacro = maxmacro
+        self.dRMS_tol = dRMS_tol
+        self.SOSCF_tol = SOSCF_tol
+        self.maxmicro = maxmicro  #microtime as member data rather than soscf return
+
+        self.Mints = psi4.MintsHelper(self.wfn.basisset())
+        self.S = np.asarray(self.Mints.ao_overlap())
+        self.T = np.asarray(self.Mints.ao_kinetic())
+        self.V = np.asarray(self.Mints.ao_potential())
+        self.H = self.T + self.V
+        del self.T, self.V, self.Mints
+ 
+        self.JK = JK_Builder() #JK_Builder probably isn't necceary as a class right now since it has one function and no member data.
+        self.Enuc = self.wfn.molecule().nuclear_repulsion_energy()
+        self.ndocc = self.wfn.nalpha()
+        self.nbf = self.S.shape[0]
+        self.nvirt = self.nbf - self.ndocc
+        self.microtime = 0.0 #SOSCF iteration data kept as class member data rather than SOSCF() returns
+        self.totmicro = 0
+
+    def diagonalize(self, F, A):
+        Fp = np.dot(A.T,F).dot(A)
+        eigval, eigvec = np.linalg.eigh(Fp)
+        C = np.dot(A,eigvec)
+        Cocc = C[:,:self.ndocc]
+        Cvirt = C[:,self.ndocc:]
+        D = np.einsum('pi,qi->pq',Cocc,Cocc)
+        return C, Cocc, Cvirt, D
```

Comment:
I would actually have ndocc as a variable here, one extra parameter and it can be very useful later on.